clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java

```diff
@@ -269,24 +297,26 @@ private <R extends MultiKeyRequestContext<K, V>, RESPONSE> void retryStreamingMu
           finalRequestCompletionFuture.completeExceptionally(throwable);
           return;
         }
-        Set<K> pendingKeys = Collections.unmodifiableSet(pendingKeysFuture.keySet());
-        R retryRequestContext =
-            requestContextConstructor.construct(pendingKeys.size(), requestContext.isPartialSuccessAllowed);
-
-        requestContext.retryContext.retryRequestContext = retryRequestContext;
-        LOGGER.debug("Retrying {} incomplete keys", retryRequestContext.numKeysInRequest);
-        // Prepare the retry context and track excluded routes on a per-partition basis
-        retryRequestContext.setRoutesForPartitionMapping(originalRequestContext.getRoutesForPartitionMapping());
-
-        streamingRequestExecutor.trigger(
-            retryRequestContext,
-            pendingKeys,
-            getStreamingCallback(
-                retryRequestContext,
-                finalRequestCompletionFuture,
-                savedException,
-                pendingKeysFuture,
-                null));
+        if (throwable != null || longTailRetryManager.isRetryAllowed()) {
```

Comment:
The long tail scenario. If the original request is taking too long then we won't see any errors yet but we want to trigger the long tail retry because it might come back fast (before the original request). 

We always send out the original/first request. If there is no retry budget and the original request never comes back then the whole request will timeout and fail (the final future).

Reply:
thanks for the explanation.